### PR TITLE
Added hello-world

### PR DESCRIPTION
**Pull request by cjdb**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+project(amcpp CXX)
+
+enable_testing()
+include(CTest)
+
+function(build file)
+   set(target "amcpp.${file}")
+   add_executable("${target}" "${file}.cpp")
+   target_compile_options("${target}" PUBLIC
+      -Wall
```
Comment (mmha) [inline CMakeLists.txt]: All these flags are GNU specific. Add them only if clang or gcc are used:
```cmake
if(${CMAKE_CXX_COMPILER_ID} MATCHES "GNU"  OR
   ${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
    target_compile_options(${target} PUBILC -Wall ...)
elseif(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
    target_compile_options(${target} PUBILC /W4 ...) # I'm not familiar with MSVC warnings apart from /W123
endif()
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      -Woverloaded-virtual
+      -Wlogical-op
+      -Wcast-qual
+      -Wold-style-cast
+      -Wnon-virtual-dtor
+      -Wodr
+      -Werror
+      -fdiagnostics-color=always
+      $<$<CONFIG:Debug>:
+         -O0
```
Comment (mmha) [inline CMakeLists.txt]: CMake adds the optimization level, `-DNDEBUG` and `-g` automatically, although one might disagree with the defaults (e. g. to `-Og` or not on `RelWithDebugInfo`). I think for the purposes of a C++ course it's okay to not override the defaults and let CMake do the work.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+         -g0
+         -DNDEBUG>
+      $<$<CONFIG:RelWithDebInfo>:
+         -O3
+         -g3
+         -DNDEBUG>)
+   if (NOT WIN32)
+      target_compile_options("${target}" PUBLIC
+         $<$<CONFIG:Debug>:
+            -fsanitize=address
```
Comment (mmha) [inline CMakeLists.txt]: I might be mistaken, but I think asan and ubsan are supported on Clang + Windows. For now I would replace `if(WIN32)` with `if(${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")` (I'm not sure about MinGW) and replace this with a proper `cxx_source_compiles` later.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            -fsanitize=undefined>
+         $<$<CONFIG:RelWithDebInfo>:
+            -fsanitize=address
+            -fsanitize=undefined>)
+   endif()
+   target_compile_definitions("${target}" PUBLIC
```
Comment (mmha) [inline CMakeLists.txt]: See above. CMake handles this already.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      $<$<CONFIG:Release>: NDEBUG>
+      $<$<CONFIG:MinSizeRel>: NDEBUG>)
+   target_compile_features("${target}" PUBLIC cxx_std_17)
+endfunction()
+
+include_directories(include)
```
Comment (mmha) [inline CMakeLists.txt]: `target_include_directories(${target} PUBLIC include)`

This directory doesn't exist yet. If this is meant for a future "common utility library", that library should be made an interface target and be linked against.

File: CMakeLists.txt (hunk 1 above)
Comment (cjdb) [reply CMakeLists.txt]: Right, this repo isn't Docker aware. This stuff is **only** compatible with GCC and Clang at the moment.

File: CMakeLists.txt (hunk 3 above)
Comment (cjdb) [reply CMakeLists.txt]: GCC also supports asan and ubsan. MinGW does not.

File: CMakeLists.txt (hunk 5 above)
Comment (cjdb) [reply CMakeLists.txt]: See the other pull request.

File: CMakeLists.txt (hunk 2 above)
Comment (cjdb) [reply CMakeLists.txt]: Thoughts on `-Og`?

**Pull request merged**